File: Problems/Dynamic-programming/Coin-Change.js
/**
 * @param {number[]} coins
 * @param {number} target
 * @return {number}
 * Time complexity: O(n.t). Iterating over target amounts and checking each 
 * of the n coins results in O(n.t).
Space complexity: O(t). The dp array requires space proportional to the
 target amount t.
 */
export default function minimumCoinsForChange(
    coins,
    target,
) {
    // Initialize a dp array with a large value (Infinity)
    const dp = new Array(target + 1).fill(Number.MAX_VALUE);

    // Base case: 0 coins are needed to make amount 0
    dp[0] = 0;

    // Iterate over each amount from 1 to the target amount
    for (let i = 1; i <= target; i++) {
        // Check each coin to see if it can contribute to the current amount
        for (const coin of coins) {
            if (i - coin >= 0) {
                dp[i] = Math.min(dp[i], dp[i - coin] + 1);
            }
        }
    }

    // If the value at dp[amount] is still Infinity, return -1 (not possible)
    // Otherwise, return the minimum number of coins needed for the target amount
    return dp[target] === Number.MAX_VALUE ? -1 : dp[target];
}

/*
Time complexity: O(ct). The recursion explores all combinations of c coins for the target
 amount t, resulting in exponential time complexity.
Space complexity: O(t). The recursive call stack grows proportionally to the target 
amount t.
*/
function minimumCoinsForChangeHelper(coins, target) {
    // Base case: If the amount is 0, no coins are needed.
    if (target === 0) {
        return 0;
    }

    let minCoins = Infinity;

    // Try each coin denomination.
    for (const coin of coins) {
        // Check if the coin denomination is not greater than the amount.
        if (coin <= target) {
            // Recursively calculate the minimum coins needed for the remaining amount.
            const remainingCoins = minimumCoinsForChangeHelper(coins, target - coin);

            // If a valid solution is found and it requires fewer coins, update minCoins.
            if (remainingCoins !== -1 && remainingCoins + 1 < minCoins) {
                minCoins = remainingCoins + 1;
            }
        }
    }

    // If no valid solution was found, return -1. Otherwise, return the minimum coins.
    return minCoins === Infinity ? -1 : minCoins;
}

export default function minimumCoinsForChange(
    coins,
    target,
) {
    return minimumCoinsForChangeHelper(coins, target);
}

File: Problems/Dynamic-programming/Coin-Change/Coin-Change.cs
using System;

// Using bottom-up approach - Time: O(n * amount)

/*
/**
 * @param {number[]} coins
 * @param {number} amount
 * @return {number}
var coinChange = function(coins, amount) {
    let n = coins.length;
    let inf = Number.MAX_VALUE;
    let dp = Array(10001).fill(inf);
    dp[0] = 0;
    for(let t = 1; t <= amount; t++) {
        for(var i = 0; i < n; i++) {
            dp[t] = Math.min(dp[t], t - coins[i] >= 0 ? 
            1 + dp[t-coins[i]] : inf);
        }
    }

    return dp[amount] == inf ? -1 : dp[amount];
};
*/

public class Solution2 {
    public int CoinChange(int[] coins, int amount) {
        int[] dp = new int[amount + 1];
        Array.Fill(dp, amount + 1);  // Initialize with a value greater than any possible minimum
        dp[0] = 0;

        for (int i = 1; i <= amount; i++) {
            foreach (int coin in coins) {
                if (i - coin >= 0) {
                    dp[i] = Math.Min(dp[i], dp[i - coin] + 1);
                }
            }
        }

        return dp[amount] > amount ? -1 : dp[amount];
    }
}

public class Solution
{
    public int CoinChange(int[] coins, int amount)
    {
        int n = coins.Length;
        int inf = 0x3f3f3f3f; // inf = 1061109567
        int[] dp = new int[10001];
        Array.Fill(dp, inf);
        dp[0] = 0;

        for (int t = 1; t <= amount; t++)
        {
            for (int i = 0; i < n; i++)
            {
                dp[t] = Math.Min(dp[t], t - coins[i] >= 0 ? 1 + dp[t - coins[i]] : inf);
            }
        }

        return dp[amount] == inf ? -1 : dp[amount];
    }
}

class Program
{
    static void Main(string[] args)
    {
        int[] coins = { 1, 2, 5 };
        int amount = 11;

        Console.WriteLine("Input: coins = [" + string.Join(", ", coins) + "], amout = " + amount);

        Solution sol = new Solution();
        int result = sol.CoinChange(coins, amount);


        Console.WriteLine("Output: "+ result);
    }
}
